### PR TITLE
Add opt-in rubric scoring + Bradley-Terry analysis, and centralize logging

### PR DESCRIPTION
**Pull request by kargibora**
## Summary
This PR introduces the **rubrics + Bradley-Terry evaluation pipeline** into OpenJury as an **opt-in extension** to the existing pairwise judge workflow.

The focus of this PR is feature addition (not behavior replacement):
- Existing pairwise evaluation remains the default
- Rubric scoring is optional
- Bradley-Terry fitting is optional
- New logging support is added for clearer outputs
- Custom rubrics can be loaded from JSON

---

### 1. Rubric-Based Evaluation (Opt-In)
Adds a structured evaluation mode where the judge scores outputs across multiple dimensions (e.g. `adherence`, `helpfulness`, `clarity`, `fluency`, etc.) instead of only giving a single pairwise preference.

#### What this enables
- Interpretable evaluation (dimension-level scores)
- Reusable built-in rubrics (`default`, `coding`, `translation`, `overall`)
- Pairwise rubric comparisons (A vs B)
- Saved rubric outputs for downstream analysis

#### Where it is available
- `openjury.generate_and_evaluate` (CLI + pipeline entrypoint)
- `openjury.evaluate.evaluate_completions(...)` (standalone evaluation path)

#### Important
Rubrics are **not forced**. The old evaluation path still works unchanged unless explicitly enabled.

---

### 2. Bradley-Terry (Feature-Based) Analysis (Opt-In)
Adds an interpretable Bradley-Terry model over rubric features.

Instead of only reporting who wins, this lets us estimate:
- Which rubric dimensions are most predictive of preference
- The relative importance of dimensions (weights)
- A fitted preference model based on rubric-score differences

#### What is saved
- BT weights per rubric dimension
- Intercept / regularization metadata

This is optional and only runs when explicitly enabled.

---

### 3. Custom Rubrics via JSON (`--rubric_json`)
Adds support for defining a rubric externally as a JSON file and using it directly in evaluation.

#### Why this matters
- No code change needed to experiment with new evaluation dimensions
- Easy to create task-specific rubrics (coding, summarization, multilingual, safety, etc.)
- Supports weighted dimensions and score anchors

#### Precedence
If both are provided:
- `--rubric_json` **overrides** `--rubric_name`

---

### 4. Centralized Logging (`openjury._logging`)
Adds a Rich-based centralized logger and results printer.

#### Benefits
- Cleaner terminal output
- Consistent logging interface (`logger.info`, `logger.warning`, ...)
- Better visibility for rubric / BT pipeline progress

This also supports future cleanup of `print()`-heavy codepaths.

---

### 5. Rubric Output Artifacts (Saved to Result Folders)
When rubric scoring is enabled, the pipeline now saves rubric and BT outputs alongside standard evaluation outputs.

#### Saved files
- `*-rubric-<rubric_name>-scores-A.csv`
- `*-rubric-<rubric_name>-scores-B.csv`
- `*-rubric-<rubric_name>-preferences.csv`
- `*-rubric-<rubric_name>-pairwise.csv`
- `*-rubric-<rubric_name>-summary.json`
- `*-rubric-<rubric_name>-bt-weights.json` (if BT is enabled)

If `--rubric_json` is used, filenames use the rubric name defined in the JSON.

---

### 6. Standalone Evaluation Parity (`evaluate.py`)
Rubric + BT support is also added to `evaluate.py` (standalone evaluation path), not only `generate_and_evaluate.py`.

This means users can:
- Evaluate existing completions with rubrics
- Fit BT without running generation
- Reuse the same judge model and result workflow

Default behavior remains unchanged unless enabled.

---

### 7. Tests (Smoke Coverage)
Adds/updates smoke tests for rubric + BT integration using lightweight stubs / monkeypatching:
- No real model required
- No network required
- Covers rubric hook + output writing flow
- Covers `rubric_json` forwarding path

---

## Backward Compatibility
This PR is designed to be safe for existing users:
- Legacy pairwise judge evaluation remains the default
- Rubrics only run when explicitly enabled
- Bradley-Terry only runs when explicitly enabled
- Existing result outputs are still generated

---

### A. Built-in Rubric + Bradley-Terry
```bash
uv run -m openjury.generate_and_evaluate \
  --dataset alpaca-eval \
  --model_A gpt4_1106_preview \
  --model_B llama-2-70b-chat-hf \
  --judge_model VLLM/Qwen/Qwen2.5-1.5B-Instruct \
  --n_instructions 10 \
  --enable_rubrics \
  --rubric_name default \
  --fit_bradley_terry \
  --bt_regularization 0.01 \
  --swap_mode fixed \
  --result_folder results_rubric_bt
```

### B. Custom Rubric JSON + Bradley-Terry
```bash
uv run -m openjury.generate_and_evaluate \
  --dataset alpaca-eval \
  --model_A gpt4_1106_preview \
  --model_B llama-2-70b-chat-hf \
  --judge_model VLLM/Qwen/Qwen2.5-1.5B-Instruct \
  --n_instructions 10 \
  --enable_rubrics \
  --rubric_name default \
  --rubric_json my_rubric.json \
  --fit_bradley_terry \
  --result_folder results_custom_rubric_bt
```

### C. Example Rubric
```json
{
  "name": "my_custom",
  "description": "Custom rubric for concise technical answers.",
  "dimensions": [
    {
      "name": "correctness",
      "description": "Is the answer technically correct?",
      "scale_min": 1,
      "scale_max": 10,
      "weight": 1.5,
      "score_references": {
        "10": "Fully correct with no meaningful errors.",
        "7": "Mostly correct with minor issues.",
        "4": "Partially correct but misses important details.",
        "1": "Incorrect or misleading."
      }
    },
    {
      "name": "clarity",
      "description": "Is the answer easy to understand and well-structured?",
      "scale_min": 1,
      "scale_max": 10,
      "weight": 1.0
    },
  ]
}
```

### CODE REVIEW

File: openjury/bradley_terry/model.py [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+        if n_dropped_nan > 0:
+            logger.warning(
+                "Dropped %d/%d rows with NaN values", n_dropped_nan, len(valid)
+            )
+        phi_A = phi_A[valid]
+        phi_B = phi_B[valid]
+        y_raw = y_raw[valid]
+
+        # Remove ties (y ≈ 0.5) — they degrade BCE fitting
+        not_tie = np.abs(y_raw - 0.5) > 0.05
```
Comment (geoalgo) [inline openjury/bradley_terry/model.py]: can we make it an option `keep_tie`? (possibly set to false by default). Since tie also contains information, ideally we would be able to exploit it at one point.

File: openjury/rubrics/__init__.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    "PairwiseRubricResult",
+    "DEFAULT_RUBRICS",
+    "RUBRIC_REGISTRY",
+    "RubricScorer",
+    "get_rubric",
+    "register_rubric",
+    "load_rubric_from_json",
+    "register_rubric_from_json",
+    "resolve_rubric",
+]
```
Comment (geoalgo) [inline openjury/rubrics/__init__.py]: We do not need those shortcuts for imports. Generally, shortcuts are useful for user facing functions but those are internals.

```suggestion
```

File: openjury/rubrics/defaults.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            ),
+        )
+    ],
+    description="Single-dimension rubric for overall quality assessment.",
+)
+
+
+# ── Registry ────────────────────────────────────────────────────────
+
+RUBRIC_REGISTRY: dict[str, Rubric] = {
```
Comment (geoalgo) [inline openjury/rubrics/defaults.py]: Can we keep it simple for a first version? We have rubrics for coding, translation for instance but those are not tested. 

Lets start simple with just a list of rubrics items that can be used like the ones in default.

File: openjury/rubrics/schema.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        instruction_index: Index linking back to the instruction.
+        model: Model that generated the completion.
+        scores: Dict mapping dimension name → score.
+        raw_judge_output: Raw text from the judge (for debugging).
+    """
+
+    instruction_index: int | str
+    model: str
+    scores: dict[str, float]
+    raw_judge_output: str = ""
```
Comment (geoalgo) [inline openjury/rubrics/schema.py]:
```suggestion
    raw_judge_output: str | None = None
```

File: openjury/rubrics/scorer.py [hunk 5]
```diff
@@ -0,0 +1,678 @@
+"""Score completions on rubric dimensions using an LLM judge.
+
+Supports two modes:
+
+- **Pairwise** (default): Judge sees both A and B side-by-side, scores each
+  on rubric dimensions, then gives an overall preference. N calls (2N with
+  swap-debiasing). Best for ranking.
```
Comment (geoalgo) [inline openjury/rubrics/scorer.py]: Why is it better for ranking?

File: openjury/rubrics/scorer.py [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+            dim.name: min(
+                dim.scale_max,
+                max(dim.scale_min, _sample_vals[i % len(_sample_vals)] - 1),
+            )
+            for i, dim in enumerate(rubric.dimensions)
+        }
+        _example_pairwise = json.dumps(
+            {"scores_A": _example, "scores_B": _example_b, "preference": "A"},
+            indent=2,
+        )
```
Comment (geoalgo) [inline openjury/rubrics/scorer.py]: Can this one moved to single function which is tested?

File: openjury/evaluate.py [hunk 7]
```diff
@@ -176,6 +203,54 @@ def get_output(df_outputs: pd.DataFrame, dataset: str, method: str):
     with open(output_folder / "results.json", "w") as f:
         json.dump(results, f)
 
+    if enable_rubrics:
```
Comment (geoalgo) [inline openjury/evaluate.py]: I do not get this, we run still run both annotations from the previous judge and then rubrics?

File: openjury/generate_and_evaluate.py [hunk 8]
```diff
@@ -223,6 +268,12 @@ def load_contexts(dataset: str) -> pd.Series:
 
 def print_results(results):
     """Print battle results in a nice formatted way"""
+    if _print_results_rich is not None:
+        try:
+            _print_results_rich(results)
+            return
+        except Exception:
+            pass
```
Comment (geoalgo) [inline openjury/generate_and_evaluate.py]:
```suggestion
```

File: openjury/bradley_terry/model.py [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+from sklearn.preprocessing import StandardScaler
+
+from openjury._logging import logger
+
+
+@dataclass
+class FeatureBradleyTerry:
+    """Bradley-Terry model with interpretable feature weights.
+
+    Wraps an sklearn estimator (default: LogisticRegression) that operates
```
[the remark this answers is not in the document]
Comment (kargibora) [reply openjury/bradley_terry/model.py]: Should it be a optional dependency or a must? I guess optional is safer right not everybody wants to do some analysis
